python_modules/dagster/dagster/components/core/decl.py

```diff
@@ -0,0 +1,310 @@
+    ComponentLoaderDecls.
+    """
+
+    def __init__(self, context: ComponentLoadContext, decls: Mapping[str, ComponentLoaderDecl]):
+        self.context = context
+        self.decls = decls
+
+    def _load_component(self) -> "CompositeComponent":
+        return CompositeComponent(
+            components={attr: decl._load_component() for attr, decl in self.decls.items()}  # noqa: SLF001
+        )
+
+
+class YamlDecl(ComponentDecl):
+    """Declaration of a single component loaded from a YAML file."""
+
+    @staticmethod
+    def from_source_tree(
+        context: ComponentLoadContext,
+        source_tree: ValueAndSourcePositionTree,
+    ) -> "YamlDecl":
+        component_file_model = _parse_and_populate_model_with_annotated_errors(
+            cls=ComponentFileModel, obj_parse_root=source_tree, obj_key_path_prefix=[]
+        )
+        return YamlDecl(
+            context=context, source_tree=source_tree, component_file_model=component_file_model
+        )
+
+    def __init__(
+        self,
+        context: ComponentLoadContext,
+        source_tree: ValueAndSourcePositionTree,
+        component_file_model: ComponentFileModel,
+    ):
+        self.context = context
+        self.source_tree = source_tree
+        self.component_file_model = component_file_model
+
+    def _load_component(self) -> "Component":
+        # find the component type
+        type_str = self.context.normalize_component_type_str(self.component_file_model.type)
+        key = PluginObjectKey.from_typename(type_str)
+        obj = load_package_object(key)
+        if not isinstance(obj, type) or not issubclass(obj, Component):
+            raise DagsterInvalidDefinitionError(
+                f"Component type {type_str} is of type {type(obj)}, but must be a subclass of dagster.Component"
+            )
+
+        context = context_with_injected_scope(
+            self.context, obj, self.component_file_model.template_vars_module
+        )
+
+        context = context.with_source_position_tree(
+            self.source_tree.source_position_tree,
+        )
+
+        model_cls = obj.get_model_cls()
+
+        # grab the attributes from the yaml file
+        if model_cls is None:
+            attributes = None
+        elif self.source_tree:
+            attributes_position_tree = self.source_tree.source_position_tree.children["attributes"]
+            with enrich_validation_errors_with_source_position(
+                attributes_position_tree, ["attributes"]
+            ):
+                attributes = TypeAdapter(model_cls).validate_python(
+                    self.component_file_model.attributes
+                )
+        else:
+            attributes = TypeAdapter(model_cls).validate_python(
+                self.component_file_model.attributes
+            )
+
+        return obj.load(attributes, context)
+
+
+class CompositeYamlDecl(ComponentDecl[CompositeYamlComponent]):
```

Comment:
I think "YamlFile" or "YamlStream" is more inline with terminology out there, with "YamlDocument" being the singular.

Reply:
updated to `YamlFileDecl`